### PR TITLE
Add VISRegLoss (VISReg paper)

### PR DESCRIPTION
**Pull request by saud5150**
Related to #1960 (PR 1 of 2).

## Description
- [ ] My change is breaking

Adds `lightly/loss/visreg_loss.py` implementing [VISReg](https://arxiv.org/abs/2606.02572) (Wu et al., 2026) per Algorithm 1, structured like `lejepa_loss.py`. Covers the open points from #1960:

1. **Distributed**: per-device slices, no `broadcast`/`all_reduce` (Section 3.2, avoiding #1920). `gather_distributed=True` optionally gathers embeddings for the statistics via the same autograd-aware path as `VICRegLoss`.
2. **Component losses**: `visreg_center/scale/shape_loss` helpers plus `forward_components()` returning a `(total, pred, scale, shape, center)` NamedTuple (per #1422). `forward()` returns a scalar.
3. **No new head**: reuses `lejepa_invariance_loss`; intended for `LeJEPAProjectionHead`.
4. **Hyperparameters**: paper's ImageNet-1K defaults (`lambda_param=0.9`, `num_slices=4096`, equal weights).

Implemented from the paper's math only (official repo is CC BY-NC). Three choices worth review: `forward(*, local_proj, global_proj)` rather than the issue's `forward(z)` sketch (the prediction term needs the global views); prediction and regularization run over **all** views per Eq. 8 / Section 3.1 (differs from `LeJEPALoss`'s local-only, pinned by a test); and Algorithm 1's `(N,K)-(N,)` broadcasting bug is fixed with an `unsqueeze`.

## Tests
- [ ] My change is covered by existing tests.
- [x] My change needs new tests.
- [x] I have added/adapted the tests accordingly.
- [x] I have manually tested the change. if_yes_describe_how

37 tests in `tests/loss/test_visreg_loss.py` (patterns from `test_lejepa_loss.py`): shape loss pinned against a naive Algorithm 1 transcription, non-vanishing gradient under collapse, stop-gradient behavior, all-views regularization scope, no-collectives on the default distributed path, and gather matching the global-batch loss. `make format`, `make static-checks`, and `make test` pass (1693 passed; 4 failures are pre-existing macOS/Python-3.9 env issues, not hit by CI).

## Documentation
- [x] I have added docstrings to all public functions/methods.
- [x] My change requires a change to the documentation ( `.rst` files).
- [x] I have updated the documentation accordingly.
- [x] The autodocs update the documentation accordingly.

`VISRegLoss` autoclass entry added to `docs/source/lightly.loss.rst` (with `:special-members: __init__`). PDF of the rendered page attached.

## Implications / comments / further issues
- PR 2 (#1960): `examples/` + `docs/source/examples/visreg.rst`, building on this loss.

### CODE REVIEW

**Comment by saud5150** [comment]
[VISRegLoss_docs_page.pdf](https://github.com/user-attachments/files/29763523/VISRegLoss_docs_page.pdf)

**Comment by gabrielfruet** [comment]
Thank you so much for the nice contribution!!!

File: tests/loss/test_visreg_loss.py [hunk 1]
```diff
@@ -0,0 +1,405 @@
...
+        # relies on it to survive fully collapsed embeddings.
+        z = torch.full((32, 16), 5.0, requires_grad=True)
+        loss = visreg_scale_loss(z)
+        loss.backward()
+        assert z.grad is not None
+        assert z.grad.isfinite().all()
+
+
+class TestVisregShapeLoss:
+    def test_matches_naive_reference(self) -> None:
```
Comment (gabrielfruet) [inline tests/loss/test_visreg_loss.py:86]: This test is really nice!! :)

File: tests/loss/test_visreg_loss.py [hunk 2]
```diff
@@ -0,0 +1,405 @@
...
+
+        torch.manual_seed(42)
+        loss = visreg_shape_loss(theta * z, num_slices=16)
+        loss.backward()
+
+        assert theta.grad is not None
+        assert theta.grad.abs() > 1e-4
+
+    def test_gaussian_scores_better_than_exponential(self) -> None:
+        torch.manual_seed(0)
```
Comment (gabrielfruet) [inline tests/loss/test_visreg_loss.py:140]: I personally like these kindas of tests. I think they show empirically that our implementation makes sense.

File: tests/loss/test_visreg_loss.py [hunk 3]
```diff
@@ -0,0 +1,405 @@
...
+        with pytest.raises(ValueError):
+            visreg_shape_loss(torch.randn(8, 4), num_slices=0)
+
+    def test_eps_must_be_positive(self) -> None:
+        with pytest.raises(ValueError):
+            visreg_shape_loss(torch.randn(8, 4), num_slices=16, eps=0.0)
+
+
+class TestVISRegLoss:
+    def test_forward(self) -> None:
```
Comment (gabrielfruet) [inline tests/loss/test_visreg_loss.py:169]: I'm missing the test_forward__cuda that we have in other losses. You can take a look at the other losses.

**Review by gabrielfruet — COMMENTED** [review]
Implementation is looking good to me :). Very nice and clean PR. I just think it's missing the `test_forward__cuda`. Thank you for the nice PR!

**Comment by gabrielfruet** [comment]
The test failure is not related to your changes FYI.

**Comment by saud5150** [comment]
Thanks for the review! Added test_forward__cuda in 03d90e3, following the other loss tests.

**Review by gabrielfruet — APPROVED** [review]
LGTM! Thank you for the nice contribution :)
